File: backend/src/main/java/com/backend/ord/domain/entities/Bank.java
package com.backend.ord.domain.entities;

import com.backend.ord.domain.entities.abstracts.EntityBase;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;

@Entity
@Table(name = "banks")
public class Bank extends EntityBase {
    @Size(max = 64)
    @NotNull
    @Column(name = "name", nullable = false, length = 64)
    private String name;

    @Size(max = 255)
    @NotNull
    @Column(name = "description", nullable = false)
    private String description;

    @NotNull
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @OnDelete(action = OnDeleteAction.CASCADE)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToOne(fetch = FetchType.LAZY)
    @OnDelete(action = OnDeleteAction.SET_NULL)
    @JoinColumn(name = "group_id")
    private BankGroup group;

    public Bank(@Size(max = 64) @NotNull String name, @Size(max = 255) @NotNull String description, @NotNull User user, BankGroup group) {
        this.name = name;
        this.description = description;
        this.user = user;
        this.group = group;
    }

    public Bank() {
    }

    public static BankBuilder builder() {
        return new BankBuilder();
    }

    public @Size(max = 64) @NotNull String getName() {
        return this.name;
    }

    public @Size(max = 255) @NotNull String getDescription() {
        return this.description;
    }

    public @NotNull User getUser() {
        return this.user;
    }

    public BankGroup getGroup() {
        return this.group;
    }

    public void setName(@Size(max = 64) @NotNull String name) {
        this.name = name;
    }

    public void setDescription(@Size(max = 255) @NotNull String description) {
        this.description = description;
    }

    public void setUser(@NotNull User user) {
        this.user = user;
    }

    public void setGroup(BankGroup group) {
        this.group = group;
    }

    public static class BankBuilder {
        private @Size(max = 64) @NotNull String name;
        private @Size(max = 255) @NotNull String description;
        private @NotNull User user;
        private BankGroup group;

        BankBuilder() {
        }

        public BankBuilder name(@Size(max = 64) @NotNull String name) {
            this.name = name;
            return this;
        }

        public BankBuilder description(@Size(max = 255) @NotNull String description) {
            this.description = description;
            return this;
        }

        public BankBuilder user(@NotNull User user) {
            this.user = user;
            return this;
        }

        public BankBuilder group(BankGroup group) {
            this.group = group;
            return this;
        }

        public Bank build() {
            return new Bank(this.name, this.description, this.user, this.group);
        }

        public String toString() {
            return "Bank.BankBuilder(name=" + this.name + ", description=" + this.description + ", user=" + this.user + ", group=" + this.group + ")";
        }
    }
}

File: backend/src/main/kotlin/com/backend/ord/features/bank/model/Bank.kt
package com.backend.ord.features.bank.model

import com.backend.ord.core.user.model.UserEntity
import com.backend.ord.domain.persistence.entities.BankGroup
import com.backend.ord.shared.models.IdentifiableUserResource
import jakarta.persistence.*
import jakarta.validation.constraints.Size
import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.OnDelete
import org.hibernate.annotations.OnDeleteAction
import org.hibernate.annotations.UpdateTimestamp
import java.time.Instant
import java.util.*

@Entity
@Table(name = "banks")
data class Bank(
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    override var id: UUID = UUID.randomUUID(),

    @field:Size(max = 64)
    @Column(name = "name", nullable = false, length = 64)
    var name: String,

    @field:Size(max = 255)
    @Column(name = "description", nullable = false)
    var description: String,

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @OnDelete(action = OnDeleteAction.CASCADE)
    @JoinColumn(name = "user_id", nullable = false)
    override var user: UserEntity,

    @ManyToOne(fetch = FetchType.LAZY)
    @OnDelete(action = OnDeleteAction.SET_NULL)
    @JoinColumn(name = "group_id")
    var bankGroup: BankGroup? = null,

    @Column(name = "group_id", nullable = true, insertable = false, updatable = false)
    var bankGroupId: UUID? = bankGroup?.id,

    @Column(name = "created_at", nullable = false, updatable = false)
    @CreationTimestamp
    var createdAt: Instant = Instant.now(),

    @Column(name = "updated_at", nullable = false)
    @UpdateTimestamp
    var updatedAt: Instant = Instant.now()
) : IdentifiableUserResource {
    @PostLoad
    fun populateBankGroupId(){
        bankGroupId = bankGroup?.id
    }
}
